src/smt/optimization_solver.cpp

```diff
@@ -46,92 +46,371 @@ OptResult OptimizationSolver::checkOpt()
-
   // we need to be in incremental mode for multiple objectives since we need to
   // push pop we need to produce models to inrement on our objective
   optChecker->setOption("incremental", "true");
   optChecker->setOption("produce-models", "true");
-
   // Move assertions from the parent solver to the subsolver
-  std::vector<Node> p_assertions = d_parent->getExpandedAssertions();
+  std::vector<Node> p_assertions = parentSMTSolver->getExpandedAssertions();
   for (const Node& e : p_assertions)
   {
     optChecker->assertFormula(e);
   }
 
-  // We need to checksat once before the optimization loop so we have a
-  // baseline value to increment
-  Result loop_r = optChecker->checkSat();
+  Result intermediateSatResult = optChecker->checkSat();
+  // Model-value of objective (used in optimization loop)
+  Node value;
+  if (intermediateSatResult.isUnknown())
+  {
+    return std::make_pair(OptResult::OPT_UNKNOWN, value);
+  }
+  if (!intermediateSatResult.isSat())
+  {
+    return std::make_pair(OptResult::OPT_UNSAT, value);
+  }
+  // asserts objective > old_value (used in optimization loop)
+  Node increment;
+  Kind incrementalOperator = kind::NULL_EXPR;
+  if (objType == ObjectiveType::OBJECTIVE_MINIMIZE)
+  {
+    // if objective is MIN, then assert optimization_target <
+    // current_model_value
+    incrementalOperator = kind::LT;
+  }
+  else if (objType == ObjectiveType::OBJECTIVE_MAXIMIZE)
+  {
+    // if objective is MAX, then assert optimization_target >
+    // current_model_value
+    incrementalOperator = kind::GT;
+  }
+  // Workhorse of linear search:
+  // This loop will keep incrmenting/decrementing the objective until unsat
+  // When unsat is hit,
+  // the optimized value is the model value just before the unsat call
+  while (intermediateSatResult.isSat())
+  {
+    value = optChecker->getValue(target);
+    Assert(!value.isNull());
+    increment = nm->mkNode(incrementalOperator, target, value);
+    optChecker->assertFormula(increment);
+    intermediateSatResult = optChecker->checkSat();
+  }
+  return std::make_pair(OptResult::OPT_OPTIMAL, value);
+}
+
+std::pair<OptResult, Node> OMTOptimizerInteger::minimize(
+    SmtEngine* parentSMTSolver, Node target)
+{
+  return this->optimize(
+      parentSMTSolver, target, ObjectiveType::OBJECTIVE_MINIMIZE);
+}
+std::pair<OptResult, Node> OMTOptimizerInteger::maximize(
+    SmtEngine* parentSMTSolver, Node target)
+{
+  return this->optimize(
+      parentSMTSolver, target, ObjectiveType::OBJECTIVE_MAXIMIZE);
+}
+
+OMTOptimizerBitVector::OMTOptimizerBitVector(bool isSigned)
+    : d_isSigned(isSigned)
+{
+}
 
-  if (loop_r.isUnknown())
+BitVector OMTOptimizerBitVector::computeAverage(const BitVector& a,
+                                                const BitVector& b,
+                                                bool isSigned)
+{
+  // computes (a + b) / 2 without overflow
+  // rounding towards -infinity: -1.5 --> -2,  1.5 --> 1
```

Comment:
enough to have in h file

Reply:
it won't bother if adding to the cpp file.